cli/util/eta.go

```diff
@@ -0,0 +1,74 @@
+package cliutil
+
+import (
+	"fmt"
+	"time"
+)
+
+// ETA implements a very simple eta calculator based on the number of remaining items. It does not
+// require knowing the work size in advance and is therefore suitable for streaming workloads and
+// also does not require that consecutive updates have a monotonically decreasing remaining value.
+type ETA struct {
+	// max number of items to keep in memory
+	maxItems int
+	// a queue of most recently updated items
+	items []item
+	// we store the last calculated ETA which we reuse if there was not change in remaining items
+	cachedETA string
+}
+
+type item struct {
+	timestamp time.Time
+	remaining int64
+}
+
+// NewETA creates a new ETA calculator of the given size
+func NewETA(maxItems int) *ETA {
+	return &ETA{
+		maxItems: maxItems,
+		items:    make([]item, 0),
+	}
+}
+
+// Update updates the ETA calculator with the remaining number of items and returns the ETA
+func (e *ETA) Update(remaining int64) string {
+	item := item{
+		timestamp: time.Now(),
+		remaining: remaining,
+	}
+
+	if len(e.items) == 0 {
+		e.items = append(e.items, item)
+		return ""
+	}
+
+	// we ignore updates with the same remaining value and just return the previous ETA
+	if e.items[len(e.items)-1].remaining == remaining {
+		return e.cachedETA
+	}
+
+	// append the item to the queue and remove the oldest item if needed
+	if len(e.items) >= e.maxItems {
+		e.items = e.items[1:]
+	}
+	e.items = append(e.items, item)
+
+	// calculate the average processing time per item in the queue
+	diffMs := e.items[len(e.items)-1].timestamp.Sub(e.items[0].timestamp).Milliseconds()
+	avg := diffMs / int64(len(e.items))
```

Comment:
Can we guarantee that we have one update per epoch? If not, I assume we need to store the number of epochs processed (or heights, I guess) as well.

Of course, heights can go backwards if we revert...

Reply:
I think this kind of works _if and only if_ we process blocks slower than once per second. In that case, each item will correspond to one block.

Otherwise, `avg` will be 1s. Or 10s/10. I.e., the time between the first item and the last (10s because we update once per second) and the number of items (10).

Reply:
@Stebalien note that we only add items if the `remaining` changes so I don't think that's an issue, or am I misunderstanding?

Reply:
Let's say remaining reduces by 5 each time. We should estimate that we're processing 5 blocks per second. However, instead, we'll estimate that we're processing 1 item per second and multiply `1 * remaining`.

Reply:
Ah now I see what you mean, and yes you are indeed correct here, will update the PR, good catch!